auxx/src/Lang/Parser.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE RecursiveDo #-}
+
+module Lang.Parser
+       ( parse
+       , ParseError(..)
+       ) where
+
+import           Universum
+
+import           Control.Applicative.Combinators.NonEmpty (sepBy1)
+import           Control.Lens                             (Getting)
+import           Data.Monoid                              (First)
+import           Text.Earley                              (Grammar, Parser, Prod, Report,
+                                                           fullParses, parser, rule,
+                                                           terminal, (<?>))
+
+import           Lang.Lexer                               (BracketSide, Token,
+                                                           getFilePath', tokenize,
+                                                           _BracketSideClosing,
+                                                           _BracketSideOpening,
+                                                           _TokenAddress,
+                                                           _TokenBlockVersion,
+                                                           _TokenFilePath, _TokenHash,
+                                                           _TokenKey, _TokenName,
+                                                           _TokenNumber,
+                                                           _TokenParenthesis,
+                                                           _TokenPublicKey,
+                                                           _TokenSemicolon,
+                                                           _TokenSoftwareVersion,
+                                                           _TokenStakeholderId,
+                                                           _TokenString)
+import           Lang.Syntax                              (Arg (..), Expr (..), Lit (..),
+                                                           ProcCall (..))
+
+tok :: Getting (First a) Token a -> Prod r e Token a
+tok p = terminal (preview p)
+
+inBrackets
+    :: Getting (First ()) Token BracketSide
+    -> Prod r e Token a
+    -> Prod r e Token a
+inBrackets p r =
+    tok (p . _BracketSideOpening) *> r <* tok (p . _BracketSideClosing)
+
+gExpr :: Grammar r (Prod r Text Token Expr)
```

Comment:
I think it would be good to have a comment with this grammar in some more common convention, because this one is hard to read for people not familiar with `Earley` library (or least me). I guess it won't take much time to write it. And it's ok to do it in another PR which adds documentation to the code (AFAIU, there will be one at some point).

Reply:
You can read everything to the right of `rule` as regular parser combinators (with an `Applicative` interface), but you don't need to worry about left recursion.